### PR TITLE
feat(datasets): Add CSVDataset to dask module

### PR DESCRIPTION
**Pull request by michaelsexton**
## Description
To add the capability for CSV datasets to be used with Kedro.

## Development notes
Copied the Parquet dataset implementation for Dask and reimplemented for CSV files. Has been tested in use of Kedro pipelines, but no test classes written yet.

## Checklist

- [X] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [x] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the relevant `RELEASE.md` file
- [x] Added tests to cover my changes

### CODE REVIEW

**Comment by astrojuanlu** [comment]
Thank you for this PR @michaelsexton ! I see the PR is in draft, do you need any input from us or not yet?

**Comment by michaelsexton** [comment]
Hi @astrojuanlu I believe I have completed all the tasks necessary in draft, however the unit tests in the checks keep failing. I am getting an `AssertionError` in the [unit-tests (ubuntu-latest, 3.10) / unit-tests](https://github.com/kedro-org/kedro-plugins/actions/runs/8433151826/job/23093621937#logs) check, however I can't replicate that locallly,

File: kedro-datasets/kedro_datasets/dask/csv_dataset.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+import dask.dataframe as dd
+import fsspec
+
+from kedro.io.core import AbstractDataset, get_protocol_and_path
+
+
+# NOTE: kedro.extras.datasets will be removed in Kedro 0.19.0.
+# Any contribution to datasets should be made in kedro-datasets
+# in kedro-plugins (https://github.com/kedro-org/kedro-plugins)
```
Comment (merelcht) [inline kedro-datasets/kedro_datasets/dask/csv_dataset.py:15]: I'm guessing this is just a copy-paste, but it's no longer relevant so you can remove it.

File: kedro-datasets/tests/dask/test_csv_dataset.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        {"Name": ["Alex", "Bob", "Clarke", "Dave"], "Age": [31, 12, 65, 29]}
+    )
+    return dd.from_pandas(df, npartitions=1)
+
+
+@pytest.fixture
+def mocked_s3_object(tmp_path, mocked_s3_bucket, dummy_dd_dataframe: dd.DataFrame):
+    """Creates test data and adds it to mocked S3 bucket."""
+    pandas_df = dummy_dd_dataframe.compute()
+    # table = pa.Table.from_pandas(pandas_df)
```
Comment (merelcht) [inline kedro-datasets/tests/dask/test_csv_dataset.py:45]: Is this comment still needed?

File: kedro-datasets/tests/dask/test_csv_dataset.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    temporary_path = tmp_path / FILE_NAME
+    pandas_df.to_csv(str(temporary_path))
+    # pq.write_table(table, str(temporary_path))
```
Comment (merelcht) [inline kedro-datasets/tests/dask/test_csv_dataset.py:48]: Same here.

**Review by merelcht — COMMENTED** [review]
I left some minor comments, but other than that this looks good. I'll have a look at the builds.

**Comment by merelcht** [comment]
@michaelsexton I've fixed several issues that weren't related to your changes, but it looks like the tests you added aren't passing currently. I can replicate this locally as well. Do you have some time to fix them?

**Comment by michaelsexton** [comment]
@merelcht Thanks, I should have some time later this week. Is it OK to discuss with you why they might failing if I continue running into them? Hopefully I'll get them all to pass anyway.

**Comment by merelcht** [comment]
> @merelcht Thanks, I should have some time later this week. Is it OK to discuss with you why they might failing if I continue running into them? Hopefully I'll get them all to pass anyway.

Absolutely! I'm happy to help get this all working.

**Comment by merelcht** [comment]
Hi @michaelsexton just checking in to see if you're still interested in finishing this PR?

**Comment by ankatiyar** [comment]
Fixed the unit tests + linting!

File: kedro-datasets/RELEASE.md [hunk 4]
```diff
@@ -10,6 +10,7 @@
 | `langchain.OpenAIEmbeddingsDataset` | A dataset for loading a OpenAIEmbeddings langchain model. | `kedro_datasets_experimental.langchain` |
 | `langchain.ChatOpenAIDataset`       | A dataset for loading a ChatOpenAI langchain model.       | `kedro_datasets_experimental.langchain` |
 * Extended preview feature to `yaml.YAMLDataset`.
+* Added new `CSVDataset` to [dask](https://www.dask.org/) module.
```
Comment (merelcht) [inline kedro-datasets/RELEASE.md:13]: Nit: we usually add new datasets in a table in the release notes, like above for the experimental dataset.

**Review by merelcht — APPROVED** [review]
Since this is a contribution to the core datasets we'll need 1/2 TSC approval according to the new contribution model (https://github.com/kedro-org/kedro-plugins/blob/main/kedro-datasets/CONTRIBUTING.md#dataset-contribution-process).

The TSC consists of 20 people today, so that's 10 approvals. 

Because this dataset adheres to the standards of a core dataset: full test coverage, docstrings, and sufficient examples, and because we already have dask datasets, I give my approval to add it as a core dataset.

**Review by datajoely — APPROVED** [review]

**Review by ankatiyar — APPROVED** [review]

**Review by DimedS — APPROVED** [review]
Thank you, @michaelsexton !

**Review by astrojuanlu — APPROVED** [review]
Thank you @michaelsexton! Let's get this in 💯

**Review by tynandebold — APPROVED** [review]

File: kedro-datasets/kedro_datasets/dask/csv_dataset.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        )
+
+    def _save(self, data: dd.DataFrame) -> None:
+        data.to_csv(self._filepath, storage_options=self.fs_args, **self._save_args)
+
+    def _exists(self) -> bool:
+        protocol = get_protocol_and_path(self._filepath)[0]
+        file_system = fsspec.filesystem(protocol=protocol, **self.fs_args)
+        files = file_system.glob(self._filepath)
+        return True if files else False
```
Comment (sbrugman) [inline kedro-datasets/kedro_datasets/dask/csv_dataset.py:125]: `return bool(files)` or `return len(files) > 0` would be cleaner in my opinion as it does not re-implement `bool`

Comment (ankatiyar) [reply kedro-datasets/kedro_datasets/dask/csv_dataset.py:125]: will fix it when we have enough approvals along with release notes and tests

**Review by idanov — APPROVED** [review]

**Review by noklam — APPROVED** [review]

**Review by SajidAlamQB — APPROVED** [review]

File: kedro-datasets/kedro_datasets/dask/csv_dataset.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+
+from copy import deepcopy
+from typing import Any
+
+import dask.dataframe as dd
+import fsspec
+from kedro.io.core import AbstractDataset, get_protocol_and_path
+
+
+class CSVDataset(AbstractDataset[dd.DataFrame, dd.DataFrame]):
```
Comment (Galileo-Galilei) [inline kedro-datasets/kedro_datasets/dask/csv_dataset.py:13]: I think I'd prefer "DaskCSVDataset" to make extra clear and avoid confusion with pandas one, but the catalog would refer to it as "dask.CSVDataset" which is already explicit, and the parquet dataset does not have this "Dask" prefix, so I am fine with letting it as is.

**Review by Galileo-Galilei — APPROVED** [review]

File: kedro-datasets/kedro_datasets/dask/csv_dataset.py (hunk 6 above)
Comment (ankatiyar) [reply kedro-datasets/kedro_datasets/dask/csv_dataset.py:13]: I think there's a precedent for this naming - eg `pandas.ParquetDataset` and `dask.ParquetDataset`, `pandas.JSONDataset`, `json.JSONDataset` and `tracking.JSONDataset`

Comment (Galileo-Galilei) [reply kedro-datasets/kedro_datasets/dask/csv_dataset.py:13]: Let's keep it consistent then 👍

**Pull request merged**